### PR TITLE
dbeaver/pro#2301 support Oracle new table partitioning

### PR DESCRIPTION
**Pull request by LonwoLonwo**
2024-01-15 17_13_04-DBeaver Ultimate 23 3 3 - NEWTABLE_11

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/edit/OracleTableManager.java [hunk 1]
```diff
@@ -98,6 +121,64 @@ protected void addObjectExtraActions(DBRProgressMonitor monitor, DBCExecutionCon
...
+            } else {
+                // Add table partitioning during creation.
+                // Parse partition keys first.
+                findKeysByColumnNames(table, partitionedBy, table.getPartitionKeys());
+                if (CommonUtils.isEmpty(table.getPartitionKeys())) {
+                    log.warn("Can't create partitioning without key columns.");
+                } else {
+                    Collection<OracleTablePartition> partitions = table.getCachedPartitions();
+                    if (CommonUtils.isEmpty(partitions)) {
+                        log.warn("Can't create partitioning without partitions. Create partitions first.");
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/edit/OracleTableManager.java]: Why warning? throw an exception

File: plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/edit/OracleTableManager.java [hunk 2]
```diff
@@ -98,6 +121,64 @@ protected void addObjectExtraActions(DBRProgressMonitor monitor, DBCExecutionCon
 
     @Override
     protected void appendTableModifiers(DBRProgressMonitor monitor, OracleTable table, NestedObjectCommand tableProps, StringBuilder ddl, boolean alter) {
+        String partitionedBy = table.getPartitionedBy(monitor);
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/edit/OracleTableManager.java]: Maybe move this into separate function

File: plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/model/OracleTablePhysical.java [hunk 3]
```diff
@@ -42,40 +44,38 @@
...
 
     //private boolean valid;
     private long rowCount;
     private Long realRowCount;
     private Object tablespace;
     private boolean partitioned;
+    private String partitionedBy;
+    private String subPartitionedBy;
     private PartitionInfo partitionInfo;
-    private PartitionCache partitionCache;
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/model/OracleTablePhysical.java:63]: Let's create optional inner caches on demand (as a general rule)

File: plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/edit/OracleTableManager.java (hunk 2 above)
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/edit/OracleTableManager.java]: Maybe it is a feature for PRO?

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/edit/OracleTableManager.java (hunk 1 above)
Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.oracle/src/org/jkiss/dbeaver/ext/oracle/edit/OracleTableManager.java]: Ok, changed.

**Review by E1izabeth — APPROVED** [review]

**Comment by uslss** [comment]
cannot create a table from schema context menu

https://github.com/dbeaver/dbeaver/assets/31996417/37c8b574-1d79-4344-a8ab-f195659ed35e

**Comment by uslss** [comment]
still reproducible 
image

also cannot drop partition and create subpartition - error message is similar

https://github.com/dbeaver/dbeaver/assets/31996417/0ac7d069-e8de-481b-a7fe-5d3e5131731b

**Comment by uslss** [comment]
image

**Comment by uslss** [comment]
error still appears

https://github.com/dbeaver/dbeaver/assets/31996417/1a02127b-27ad-4719-a5c6-5810c87413f1

**Comment by LonwoLonwo** [comment]
Fixed

**Comment by uslss** [comment]
verified

**Pull request merged**